features.js

```diff
@@ -186,65 +186,52 @@ function getCodec(peerConnectionLog, kind, direction) {
-                        break;
-                    }
-                } else {
-                    var tracks = peerConnectionLog[i].value.split(' ', 2);
-                    tracks.shift();
-                    tracks = tracks[0].split(',');
-                    for (var j = 0; j < tracks.length; j++) {
-                        if (tracks[j].split(':')[0] === kind) {
-                            trackId = tracks[j].split(':')[1];
-                            break;
-                        }
+// extracts a Map with all local and remote audio/video tracks.
+function extractTracks(peerConnectionLog) {
+    const tracks = new Map();
+    for (let i = 0; i < peerConnectionLog.length; i++) {
+        const {type, value} = peerConnectionLog[i];
+        if (type === 'addStream') {
+            const [streamId, _] = value.split(' ');
+            _.split(',').forEach(id => {
+                const [kind, trackId] = id.split(':');
+                tracks.set(trackId, {kind, streamId, direction: 'send', stats: []});
+            });
+        } else if (type === 'addTrack') {
+            const [kind, trackId] = value.split(' ')[0].split(':');
+            const streamId = value.split(' ')[1].split(':')[1];
+            tracks.set(trackId, {kind, streamId, direction: 'send', stats: []});
+        } else if (type === 'ontrack') {
+            const [kind, trackId] = value.split(' ')[0].split(':');
+            const streamId = value.split(' ')[1].split(':')[1];
+            tracks.set(trackId, {kind, streamId, direction: 'recv', stats: []});
+        } else if (type === 'getStats') {
+            Object.keys(value).forEach(id => {
+                const report = value[id];
+                if (report.type === 'ssrc') {
+                    const {trackIdentifier} =  report;
+                    if (tracks.has(trackIdentifier)) {
+                        report.timestamp = peerConnectionLog[i].time;
+                        tracks.get(trackIdentifier).stats.push(report);
+                    } else if (trackIdentifier !== undefined) {
+                        console.log('NO ONTRACK FOR', trackIdentifier, report.ssrc);
                     }
                 }
-            }
+            });
         }
-        if (trackId) break;
     }
-    if (!trackId) return []; // bail out
+    return tracks;
+}
 
-    // search for signs of that track
-    for (; i < peerConnectionLog.length; i++) {
-        if (trackId && peerConnectionLog[i].type === 'getStats') {
-            var statsReport = peerConnectionLog[i].value;
-            Object.keys(statsReport).forEach(id => {
-                var report = statsReport[id];
-                if (report.type === 'ssrc' && report.trackIdentifier === trackId) {
-                    report.timestamp = peerConnectionLog[i].time;
-                    reports.push(report);
-                }
-            });
+// extract a local/remote audio or video track.
+function extractTrack(peerConnectionLog, kind, direction) {
```

Comment:
`extractFirstTrackStats`. ?

Reply:
i'd rather not change this

Reply:
Ok, but it doesn't extract a track, but its stats :)